sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/Range.scala

```
@@ -0,0 +1,75 @@
+import org.apache.spark.sql.types.{BooleanType, DateType, TimestampType, _}
+
+
+/** Value range of a column. */
+trait Range
+
+/** For simplicity we use decimal to unify operations of numeric ranges. */
+case class NumericRange(min: JDecimal, max: JDecimal) extends Range
+
+/**
+ * This version of Spark does not have min/max for binary/string types, we define their default
+ * behaviors by this class.
+ */
+class DefaultRange extends Range
+
+/** This is for columns with only null values. */
+class NullRange extends Range
+
+object Range {
+  def apply(min: Option[Any], max: Option[Any], dataType: DataType): Range = dataType match {
+    case StringType | BinaryType => new DefaultRange()
+    case _ if min.isEmpty || max.isEmpty => new NullRange()
+    case _ => toNumericRange(min.get, max.get, dataType)
+  }
+
+  /**
+   * For simplicity we use decimal to unify operations of numeric types, the two methods below
+   * are the contract of conversion.
+   */
+  private def toNumericRange(min: Any, max: Any, dataType: DataType): NumericRange = {
+    dataType match {
+      case _: NumericType =>
+        NumericRange(new JDecimal(min.toString), new JDecimal(max.toString))
+      case BooleanType =>
+        val min1 = if (min.asInstanceOf[Boolean]) 1 else 0
+        val max1 = if (max.asInstanceOf[Boolean]) 1 else 0
+        NumericRange(new JDecimal(min1), new JDecimal(max1))
+      case DateType =>
+        val min1 = DateTimeUtils.fromJavaDate(min.asInstanceOf[Date])
+        val max1 = DateTimeUtils.fromJavaDate(max.asInstanceOf[Date])
+        NumericRange(new JDecimal(min1), new JDecimal(max1))
+      case TimestampType =>
+        val min1 = DateTimeUtils.fromJavaTimestamp(min.asInstanceOf[Timestamp])
+        val max1 = DateTimeUtils.fromJavaTimestamp(max.asInstanceOf[Timestamp])
+        NumericRange(new JDecimal(min1), new JDecimal(max1))
+      case _ =>
+        throw new AnalysisException(s"Type $dataType is not castable to numeric in estimation.")
```

when we get here, is it an error in spark? if yes, we should probably throw UnsupportedOperationEception